forest/benchmarking/stratified_experiment.py

```diff
@@ -0,0 +1,334 @@
+    """
+    Takes in StratifiedExperiments and simultaneously runs individual layers of equal depth
+    across separate experiments with allowed_parallel_types.
+
+    First, we group all the Layers that share a depth. Within one such group, we then group all
+    of the Components that act on disjoint sets of qubits and so can be run simultaneously; this
+    is done irrespective of experiment type
+
+    For each of these groups we form a single simultaneous program that is run on the qc
+    and randomly partition measurement settings to be measured simultaneously on this program
+    until all of the required measurements for each layer are taken (note this might mean
+    that the sequences of layers with fewer measurements are simply run without being
+    measured for some runs).
+
+    :param qc:
+    :param experiments:
+    :param num_shots:
+    :param parallelize_layers: if true, layers of equal depth will be run in parallel
+        wherever possible. Layers may be run in parallel if they act on disjoint sets of
+        qubits. You may require that grouping of experiments only occur for certain types; see
+        param allowed_parallel_types and method _group_by_depth
+    :param allowed_parallel_types: by default, when parallelize_layers is set to true only
+        layers in experiments of the same type will ever possibly be run in parallel. If
+        allowed_parallel_types is specified then layers whose experiment types share an inner
+        list may be run in parallel. Note that running a T1/T2 experiment in parallel with any other
+        type may affect results since measurement of all qubits occurs at the end of each program;
+        if the non T1/T2 experiment sequences take more time than the DELAY time specified in T1/T2,
+        those results will not match the intended delay time.
+    :param randomize_layers_within_depth: shuffle layers within each depth before data for that
+        depth is collected
+    :param randomize_all_layers: shuffle all of the layers before they are run, regardless of depth
+    :return: Currently mutates the input StratifiedExperiments.
+    """
+    if isinstance(experiments, StratifiedExperiment):
+        experiments = [experiments]
+
+    # make a copy of each experiment
+    experiments = [deepcopy(expt) for expt in experiments]
+
+    if parallelize_layers:
+        # by default only layers of the same expt type will possibly be parallelized.
+        expts_by_type_group = _group_allowed_types(experiments, allowed_parallel_types)
+    else:
+        # each layer of each experiment will be run iteratively
+        expts_by_type_group = [[expt] for expt in experiments]
+
+    # for each depth get a list of groups of layers from experiment types that can be parallelized
+    parallelizable_groups_by_depth = _group_by_depth(expts_by_type_group)
+    # TODO: any reason to supply option for not randomizing order of depths?
```

Comment:
We should at least have an option to add a seed for any randomization -- it is nice to have a way to reproduce behavior (if it helps to uncover or demonstrate a bug).